### PR TITLE
Support for Debug Codes - Dnnn

### PR DESCRIPTION
**Pull request by rhapsodyv**
### Description

While working with the TFT refactoring, I had to do a lot of tests with the TFT init code. But is very annoying to flash and re-flash to do simple tests. 
I wanted a way to talk to Marlin, and it execute my custom code.

So I created a support for a custom gcode that is used only for devs.

I named it M9999. The code of M9999 is what the dev need, when he needs it. 

This way, I could create my own test code, and do a lot of testing sending commands to marlin, without need to flash every time.

When enable `MARLIN_DEV_MODE` and `DEV_CUSTOM_GCODE`, the developer can write custom code in the file: `Marlin/src/gcode/dev_custom_gcode.h`.

After the initial skeleton, I added this file in the `.gitignore`, so it won't be committed by mistake.

### Benefits

Easier to test marlin when developing!

### Configurations

Enable:
```
DEV_CUSTOM_GCODE
MARLIN_DEV_MODE
```

### CODE REVIEW

**Comment by thinkyhead** [comment]
Please demonstrate its usage.

**Comment by rhapsodyv** [comment]
I need to test all those variations in the TFT refactoring:
90 no mirror
90 + mirror horizontal
90 + mirror vertical
180 no mirror
180 + mirror horizontal
180 + mirror vertical
270 no mirror
270 + mirror horizontal
270 + mirror vertical

Each one is a combination of three operations: invert X, invert Y and swap X and Y (all are bit flags). It's very un-intuitive and confusing.

90 seems to be: swap XY + invert X. 
270 seems to be swap XY + invert Y.

Its really annoying and tedious to: set some flags, build, copy to sd, flash, start, see the results. Change the flags, build, copy to sd, flash, start, see the results... over and over again, until a got everything right with 3 different boards and displays!!

But how to test each combination in a simple and iterative way?

To solve this problem, I created this special gcode, that allow me to write custom code and call it by serial with custom parameters, as I need.

In my case, the code of that M9999 is:
```
  #include "../lcd/extui/lib/mks_ui/SPI_TFT.h"
  void GcodeSuite::M9999() {
    parser.seen('V');
    const uint16_t orientation = parser.value_int();
    SERIAL_ECHOLNPAIR("New Orientation: ", orientation);
    SPI_TFT.tftio.reInitTFT(orientation);
  }
```

So, I can send the tests using serial:

To test 90, I send: `M9999 V32`  (32 is the bit flag result of swap XY | invert Y), The screen updates instantly, so I can see if my 90 is right or wrong. If is wrong, I send another bit flag combination, like `M9999 V224`, and so on...

I can test all operations and bit flags, with a single build, and a simple code to test.

Sometime in the future, I may need to test some XPT calibration values... And I can do the same, sending values and testing, without any build or any screen.

Or to test the best position for screen elements...

And so on... what the developer need to test in realtime, it could be done with this custom gcode, to avoid tedius change-build-test loop.

**Comment by rhapsodyv** [comment]
We can use Debug GCodes `D`, instead of `M`... 

https://reprap.org/wiki/G-code#D:_Debug_codes

**Comment by thinkyhead** [comment]
Prusa firmware is a beast of its own, so adding support for its family of D-codes is not exactly a trivial operation. It will require some extension of the G-code parser and other hooks that aren't yet in Marlin. But, you can use the general idea if you just want to add some debug hooks for your own usage.

**Comment by rhapsodyv** [comment]
Ok, but the general idea of having a custom gcode with a "playground" for the developer remains, even when using the Dxxx.

We may not need the new config option, but we do need a custom code and that custom header or cpp gitignored. 

This way the dev can play as he needs, without worrying about committing or staging changes by mistakes. He doesn't need to all the time stash/unstash changes too.

For example: my code to test tft will never be committed, as it's too specific for tests, and after my dev is done is completely useless.

**Comment by sjasonsmith** [comment]
Where does this stand? I needed to create a command that would trigger a watchdog reset, and adding a temporary D command was a very convenient way to do it.

> This way the dev can play as he needs, without worrying about committing or staging changes by mistakes. He doesn't need to all the time stash/unstash changes too.

I'm not very worried about this. It is easy to add a custom D command, and since they are in their own file it should be obvious if they are accidentally appearing in PRs. I dislike the idea of adding code files to `.gitignore`, because I want to commit my test code to my own branches, even if I don't plan to push them upstream.

File: Marlin/src/gcode/parser.h [hunk 1]
```diff
@@ -401,6 +410,25 @@ class GCodeParser {
...
+  #if ENABLED(MARLIN_DEV_MODE)
+
+    static inline uint8_t* hex_adr_val(const char c, uint8_t * const dval=nullptr) {
+      if (!seen(c) || *value_ptr != 'x') return dval;
+      uint8_t *out = nullptr;
+      for (char *vp = value_ptr + 1; HEXCHR(*vp) >= 0; vp++)
+        out = (uint8_t*)((uintptr_t(out) << 8) | HEXCHR(*vp));
+      return out;
+    }
+
```
Comment (thinkyhead) [inline Marlin/src/gcode/parser.h:422]: These two functions may be redundant. See the `hex_print.*` files for the existing functions and their usage.

**Comment by ellensp** [comment]
code presumes existence of eeprom (or emulation) and fails to compile with
Marlin\src\gcode\gcode_d.cpp: In static member function 'static void GcodeSuite::D(int16_t)':
Marlin\src\gcode\gcode_d.cpp:92:45: error: 'E2END' was not declared in this scope

**Comment by rhapsodyv** [comment]
> code presumes existence of eeprom (or emulation) and fails to compile with
> Marlin\src\gcode\gcode_d.cpp: In static member function 'static void GcodeSuite::D(int16_t)':
> Marlin\src\gcode\gcode_d.cpp:92:45: error: 'E2END' was not declared in this scope

Yes, it presumes... I will add some #ifdefs, because a lot of those D codes are to working with EEPROM.

**Comment by sjasonsmith** [comment]
My thought is that if people are going to enable debug mode, they should be capable of debugging problems with debug mode 😉

**Comment by ellensp** [comment]
wishful thinking...

**Pull request merged**